File: app/src/main/java/com/github/zawadz88/materialpopupmenu/internal/SectionedRecyclerViewAdapter.java
package com.github.zawadz88.materialpopupmenu.internal;
/*
 * Copyright (C) 2015 Tomás Ruiz-López.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import android.support.v7.widget.RecyclerView;
import android.view.ViewGroup;

/**
 * An extension to RecyclerView.Adapter to provide sections with headers to a
 * RecyclerView. Each section can have an arbitrary number of items.
 * <p>
 * Modified version of <a href="https://github.com/truizlop/SectionedRecyclerView">SectionedRecyclerView</a> adapter without the footers
 * and with optional section headers.
 *
 * @param <H>  Class extending RecyclerView.ViewHolder to hold and bind the header view
 * @param <VH> Class extending RecyclerView.ViewHolder to hold and bind the items view
 */
public abstract class SectionedRecyclerViewAdapter<H extends RecyclerView.ViewHolder,
        VH extends RecyclerView.ViewHolder>
        extends RecyclerView.Adapter<RecyclerView.ViewHolder> {

    private static final int TYPE_SECTION_HEADER = -1;
    private static final int TYPE_ITEM = -2;

    private int[] sectionForPosition = null;
    private int[] positionWithinSection = null;
    private boolean[] isHeader = null;
    private int count = 0;

    /**
     * Returns the sum of number of items for each section plus headers and footers if they
     * are provided.
     */
    @Override
    public int getItemCount() {
        return count;
    }

    public void setupIndices() {
        count = countItems();
        allocateAuxiliaryArrays(count);
        precomputeIndices();
    }

    private int countItems() {
        int count = 0;
        int sections = getSectionCount();

        for (int i = 0; i < sections; i++) {
            count += getItemCountForSection(i) + (hasHeaderInSection(i) ? 1 : 0);
        }
        return count;
    }

    private void precomputeIndices() {
        int sections = getSectionCount();
        int index = 0;

        for (int i = 0; i < sections; i++) {
            if (hasHeaderInSection(i)) {
                setPrecomputedItem(index, true, i, 0);
                index++;
            }

            for (int j = 0; j < getItemCountForSection(i); j++) {
                setPrecomputedItem(index, false, i, j);
                index++;
            }
        }
    }

    private void allocateAuxiliaryArrays(int count) {
        sectionForPosition = new int[count];
        positionWithinSection = new int[count];
        isHeader = new boolean[count];
    }

    private void setPrecomputedItem(int index, boolean isHeader, int section, int position) {
        this.isHeader[index] = isHeader;
        sectionForPosition[index] = section;
        positionWithinSection[index] = position;
    }

    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        RecyclerView.ViewHolder viewHolder;

        if (isSectionHeaderViewType(viewType)) {
            viewHolder = onCreateSectionHeaderViewHolder(parent, viewType);
        } else {
            viewHolder = onCreateItemViewHolder(parent, viewType);
        }

        return viewHolder;
    }

    @Override
    public void onBindViewHolder(RecyclerView.ViewHolder holder, int position) {
        int section = sectionForPosition[position];
        int index = positionWithinSection[position];

        if (isSectionHeaderPosition(position)) {
            onBindSectionHeaderViewHolder((H) holder, section);
        } else {
            onBindItemViewHolder((VH) holder, section, index);
        }

    }

    @Override
    public int getItemViewType(int position) {

        if (sectionForPosition == null) {
            setupIndices();
        }

        int section = sectionForPosition[position];
        int index = positionWithinSection[position];

        if (isSectionHeaderPosition(position)) {
            return getSectionHeaderViewType(section);
        } else {
            return getSectionItemViewType(section, index);
        }

    }

    protected int getSectionHeaderViewType(int section) {
        return TYPE_SECTION_HEADER;
    }

    protected int getSectionItemViewType(int section, int position) {
        return TYPE_ITEM;
    }

    /**
     * Returns true if the argument position corresponds to a header
     */
    public boolean isSectionHeaderPosition(int position) {
        if (isHeader == null) {
            setupIndices();
        }
        return isHeader[position];
    }

    protected boolean isSectionHeaderViewType(int viewType) {
        return viewType == TYPE_SECTION_HEADER;
    }

    /**
     * Returns true if a given section should have a header
     */
    protected abstract boolean hasHeaderInSection(int sectionPosition);

    /**
     * Returns the number of sections in the RecyclerView
     */
    protected abstract int getSectionCount();

    /**
     * Returns the number of items for a given section
     */
    protected abstract int getItemCountForSection(int section);

    /**
     * Creates a ViewHolder of class H for a Header
     */
    protected abstract H onCreateSectionHeaderViewHolder(ViewGroup parent, int viewType);

    /**
     * Creates a ViewHolder of class VH for an Item
     */
    protected abstract VH onCreateItemViewHolder(ViewGroup parent, int viewType);

    /**
     * Binds data to the header view of a given section
     */
    protected abstract void onBindSectionHeaderViewHolder(H holder, int sectionPosition);

    /**
     * Binds data to the item view for a given position within a section
     */
    protected abstract void onBindItemViewHolder(VH holder, int section, int position);

}

File: material-popup-menu/src/main/java/com/github/zawadz88/materialpopupmenu/internal/SectionedRecyclerViewAdapter.kt
package com.github.zawadz88.materialpopupmenu.internal

/*
 * Copyright (C) 2015 Tomás Ruiz-López.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView

/**
 * An extension to RecyclerView.Adapter to provide sections with headers to a
 * RecyclerView. Each section can have an arbitrary number of items.
 *
 *
 * Modified version of [SectionedRecyclerView](https://github.com/truizlop/SectionedRecyclerView) adapter without the footers
 * and converted to Kotlin.
 *
 * @param H  Class extending RecyclerView.ViewHolder to hold and bind the header view
 * @param VH Class extending RecyclerView.ViewHolder to hold and bind the items view
 * */
internal abstract class SectionedRecyclerViewAdapter<H : RecyclerView.ViewHolder, VH : RecyclerView.ViewHolder> : RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    private var sectionForPosition: IntArray? = null
    private var positionWithinSection: IntArray? = null
    private var isHeader: BooleanArray? = null
    private var count = 0

    /**
     * Returns the sum of number of items for each section plus headers and footers if they
     * are provided.
     */
    override fun getItemCount(): Int {
        return count
    }

    fun setupIndices() {
        count = countItems()
        allocateAuxiliaryArrays(count)
        precomputeIndices()
    }

    private fun countItems(): Int {
        val sections = sectionCount

        return (0 until sections).sumBy { getItemCountForSection(it) + 1 }
    }

    private fun precomputeIndices() {
        val sections = sectionCount
        var index = 0

        for (i in 0 until sections) {
            setPrecomputedItem(index, true, i, 0)
            index++


            for (j in 0 until getItemCountForSection(i)) {
                setPrecomputedItem(index, false, i, j)
                index++
            }
        }
    }

    private fun allocateAuxiliaryArrays(count: Int) {
        sectionForPosition = IntArray(count)
        positionWithinSection = IntArray(count)
        isHeader = BooleanArray(count)
    }

    private fun setPrecomputedItem(index: Int, isHeader: Boolean, section: Int, position: Int) {
        this.isHeader?.set(index, isHeader)
        sectionForPosition?.set(index, section)
        positionWithinSection?.set(index, position)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        return if (isSectionHeaderViewType(viewType)) {
            onCreateSectionHeaderViewHolder(parent, viewType)
        } else {
            onCreateItemViewHolder(parent, viewType)
        }
    }

    @Suppress("UNCHECKED_CAST")
    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        val section = sectionForPosition!![position]
        val index = positionWithinSection!![position]
        if (isSectionHeaderPosition(position)) {
            onBindSectionHeaderViewHolder(holder as H, section)
        } else {
            onBindItemViewHolder(holder as VH, section, index)
        }
    }

    override fun getItemViewType(position: Int): Int {

        if (sectionForPosition == null) {
            setupIndices()
        }

        val section = sectionForPosition!![position]
        val index = positionWithinSection!![position]

        return if (isSectionHeaderPosition(position)) {
            getSectionHeaderViewType(section)
        } else {
            getSectionItemViewType(section, index)
        }
    }

    @Suppress("UNUSED_PARAMETER")
    private fun getSectionHeaderViewType(section: Int): Int {
        return TYPE_SECTION_HEADER
    }

    @Suppress("UNUSED_PARAMETER")
    protected open fun getSectionItemViewType(section: Int, position: Int): Int {
        return TYPE_ITEM
    }

    /**
     * Returns true if the argument position corresponds to a header
     */
    private fun isSectionHeaderPosition(position: Int): Boolean {
        if (isHeader == null) {
            setupIndices()
        }
        return isHeader!![position]
    }

    private fun isSectionHeaderViewType(viewType: Int): Boolean {
        return viewType == TYPE_SECTION_HEADER
    }

    /**
     * Returns the number of sections in the RecyclerView
     */
    protected abstract val sectionCount: Int

    /**
     * Returns the number of items for a given section
     */
    protected abstract fun getItemCountForSection(section: Int): Int

    /**
     * Creates a ViewHolder of class H for a Header
     */
    protected abstract fun onCreateSectionHeaderViewHolder(parent: ViewGroup, viewType: Int): H

    /**
     * Creates a ViewHolder of class VH for an Item
     */
    protected abstract fun onCreateItemViewHolder(parent: ViewGroup, viewType: Int): VH

    /**
     * Binds data to the header view of a given section
     */
    protected abstract fun onBindSectionHeaderViewHolder(holder: H, sectionPosition: Int)

    /**
     * Binds data to the item view for a given position within a section
     */
    protected abstract fun onBindItemViewHolder(holder: VH, section: Int, position: Int)

    companion object {

        internal const val TYPE_SECTION_HEADER = -1
        internal const val TYPE_ITEM = -2
    }
}
